### PR TITLE
Inspect lazy result

### PR DESCRIPTION
**Pull request by steakknife**
Show lazy results in logs and stack traces.

Signed-off-by: Barry Allard <ball@meta.com>

## Description

`  path #<Chef::DelayedEvaluator:0x012345812384 path/to/file>` is unhelpful in resource logs and stack traces. This change fixes that. It assumes dumb, non-idempotent stuff doesn't happen in the lazy evaluation runtime phase. This will hopefully go away when `unified_mode` is possible to use with our API cookbook model.

`  path lazy { (evaluates to) "C:\Program Files\Foobar\Widgets.exe" }`

`  path lazy { (evaluates to) "/home/foo/.your-magic-file" }`

are much nicer.

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Chore (non-breaking change that does not add functionality or fix an issue)

## Checklist:
- [x] I have read the **CONTRIBUTING** document.
- [x] I have run the pre-merge tests locally and they pass.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] All commits have been signed-off for [the Developer Certificate of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

### CODE REVIEW

**Review by steakknife — COMMENTED** [review]
It would be nice(tm) to hacky cache the results of a previous `#call(*)` but it's a frozen object in some environments. It's more useful to show the value by call()'ing again than to consider edge-cases with non-idempotent side-effects.

**Review by jaymzh — APPROVED** [review]
This seems sane to me. I don't know much about the delayed evaluator, but if the tests pass I'm cool with this.

**Comment by tpowell-progress** [comment]
@steakknife can you clean up the linting errors and also re-sign the DCO @dafyddcrosby is looking into this to see if there a DCO issue as well.

File: spec/unit/delayed_evaluator_spec.rb [hunk 1]
```diff
@@ -0,0 +1,39 @@
+#
+# Author:: Adam Jacob (<adam@chef.io>)
```
Comment (marcparadise) [inline spec/unit/delayed_evaluator_spec.rb:2]: Please remove these Author entries from the header

File: spec/unit/delayed_evaluator_spec.rb [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+# Author:: Christopher Walters (<cw@chef.io>)
+# Author:: Tim Hinderliter (<tim@chef.io>)
+# Author:: Seth Chisamore (<schisamo@chef.io>)
```
Comment (marcparadise) [inline spec/unit/delayed_evaluator_spec.rb:5]:
```suggestion
```

**Review by jaymzh — CHANGES_REQUESTED** [review]
You still have lint issues, sorry. If you can fix your trailing white space we'll be happy to merge. Thanks!

**Comment by jaymzh** [comment]
@dafyddcrosby  is taking this on

**Comment by dafyddcrosby** [comment]
I've opened #13902 with the cosmetic fix to get this committed

**Comment by jaymzh** [comment]
Merged in https://github.com/chef/chef/pull/13902
